src/reflection.ts

```diff
@@ -697,4 +728,172 @@ export class Parser {
+
+    const unionDeserializers = this.createUnionDeserializers(fieldType, readDefaults);
+
+    return (table: Table) => {
+      const discriminators = readDiscriminators(table);
+      if (!discriminators) {
+        throw new Error(
+          `Malformed message: missing vector discriminators field: ${discriminatorField.name()}`,
+        );
+      }
+
+      const tables = vectorLambda(table);
+      if (tables == null) {
+        throw new Error(`Malformed message: missing vector table for field: ${field.name()}`);
+      }
+
+      if (discriminators.length !== tables.length) {
+        throw new Error(
+          `malformed message: ${field.name()} length != ${discriminatorField.name()} length`,
+        );
+      }
+
+      const result = [];
+      for (let idx = 0; idx < tables.length; ++idx) {
+        const discriminator = discriminators[idx];
+        if (typeof discriminator !== "number") {
+          throw new Error(
+            `Malformed union discriminator value is not a number in field: ${discriminatorField.name()}`,
+          );
+        }
+
+        // Skip NONE
+        if (discriminator < 0) {
+          continue;
+        }
+
+        const deserializer = unionDeserializers.get(discriminator);
+        if (!deserializer) {
+          throw new Error(
+            `Malformed message: unknown union type '${discriminator}' in field ${discriminatorField.name()}`,
+          );
+        }
+
+        const subTable = tables[idx];
+        if (!subTable) {
+          throw new Error(`Malformed message missing table at ${field.name()} positon ${idx}`);
+        }
+        result.push(deserializer(subTable));
+      }
+
+      return result;
+    };
+  }
+
+  readUnionLambda(
+    field: reflection.Field,
+    discriminatorField: reflection.Field,
+    readDefaults: boolean,
+  ): (table: Table) => Record<string, any> | undefined {
+    const fieldType = field.type();
+    if (fieldType == null) {
+      throw new Error(`Malformed schema: "type" field of '${field.name()}' not populated.`);
+    }
+
+    const discriminatorfieldType = discriminatorField.type();
+    if (discriminatorfieldType === null) {
+      throw new Error('Malformed schema: "type" field of Field not populated.');
+    }
+
+    if (!isScalar(discriminatorfieldType.baseType())) {
+      throw new Error(`Malformed schema: union discriminator field is not a scalar`);
+    }
+
+    const parseDiscriminator = this.readScalarLambda(
+      discriminatorField,
+      discriminatorfieldType.index(),
+      readDefaults,
+    );
+
+    const unionDeserializers = this.createUnionDeserializers(fieldType, readDefaults);
+
+    // Unions can only be formed from tables so we know our union field will point to a table
+    const rawLambda = this.readTableLambda(field, fieldType.index());
+
+    return (table: Table) => {
+      const discriminatorValue = parseDiscriminator(table);
+      // NONE becomes undefined when parsed
+      if (discriminatorValue == undefined) {
```

Comment:
People always tell me to always use `===`. Should I be asking for that here?

Reply:
Hah. Amusingly IMO no. In our app codebase we require the use of `===` for all comparisons _except_ those for `undefined`. The reason for this is to also check for `null`. So by using `== undefined` you get a true condition if the value is `undefined` _or_ `null` which is what you want in most cases. So this is a habit formed from our codebase and lint rules.